src/waltham/waltham-connection.c

```diff
@@ -420,3 +410,136 @@ wth_connection_assert_side(struct wth_connection *conn,
+	display_error,
+	display_delete_id,
+	display_server_version
+};
+
+WTH_EXPORT struct wthp_registry *
+wth_connection_get_registry(struct wth_connection *conn)
+{
+	ASSERT_CLIENT_SIDE(conn);
+
+	wth_display_set_listener(conn->display, &display_listener, conn);
+	return wth_display_get_registry(conn->display);
+}
+
+WTH_EXPORT struct wthp_callback *
+wth_connection_sync(struct wth_connection *conn)
+{
+	ASSERT_CLIENT_SIDE(conn);
+
+	return wth_display_sync(conn->display);
+}
+
+/* END wthp_display client implementation */
+
+/* BEGIN wthp_display server implementation */
+
+/* TODO: these three declaration are copied from waltham-server.h, which can’t
+ * be included together with waltham-client.h.  Find a way to make it possible
+ * to include them both, or split this server implementation in another file.
+ */
+
+struct wth_display_interface {
+	void (*client_version) (struct wth_display * wth_display, uint32_t client_version);
+	void (*sync) (struct wth_display * wth_display, struct wthp_callback * callback);
+	void (*get_registry) (struct wth_display * wth_display, struct wthp_registry * registry);
+};
+
+void
+wthp_callback_send_done (struct wthp_callback * wthp_callback, uint32_t callback_data);
+
+void
+wth_display_send_error (struct wth_display * wth_display, struct wth_object * object_id, uint32_t code, const char * message);
+
+static void
+display_handle_client_version(struct wth_display *wth_display,
+			      uint32_t client_version)
+{
+	wth_object_post_error((struct wth_object *)wth_display, 0,
+			      "unimplemented: %s", __func__);
+}
+
+static void
+display_handle_sync(struct wth_display *wth_display,
+                    struct wthp_callback *callback)
+{
+	wth_debug("Client requested wth_display.sync");
+	wthp_callback_send_done(callback, 0);
+	wthp_callback_free(callback);
+}
+
+static void
+display_handle_get_registry(struct wth_display *wth_display,
+			    struct wthp_registry *registry)
+{
+	struct wth_object *disp_object = (struct wth_object *)wth_display;
+	void *user_pointer = wth_object_get_user_data(disp_object);
+	disp_object->connection->registry_callback(registry, user_pointer);
+}
+
+static const struct wth_display_interface display_implementation = {
+	display_handle_client_version,
+	display_handle_sync,
+	display_handle_get_registry
+};
+
+WTH_EXPORT void
+wth_connection_set_registry_callback(struct wth_connection *conn,
+                                     void (*callback)(struct wthp_registry *, void *),
+                                     void *user_pointer)
+{
+	ASSERT_SERVER_SIDE(conn);
+
+	wth_object_set_listener((struct wth_object *)conn->display, (void (**)(void))&display_implementation, user_pointer);
+	conn->registry_callback = callback;
+}
+
+WTH_EXPORT void
+wth_connection_send_error(struct wth_connection *conn, struct wth_object *obj,
+                          uint32_t code, const char *str)
+{
+	wth_display_send_error(conn->display, obj, code, str);
```

Comment:
There is something with these error sending functions that I can't quite wrap my head around.
When a server sends an error, it should mark the connection as failed somehow.
Hmm, should this function be not exported?
I would think the public API should be `wth_connection_post_display_error()` for errors on the display itself and `wth_object_post_error()` for custom errors on a specific object. The below implementation for the former is correct.

Reply:
The issue was a circular dependency between `wth_connection`, `wth_object` and `wth_display`, with `wth_object_post_error()` sending both an error on the `wth_display` and a protocol error on `wth_connection`.

This `wth_connection_send_error()` function is here only to prevent having to export the `wth_display` object to each `wth_object`.

Reply:
I don't see what the problem with exposing internal functions to other internal `.c` files is. Do not export the functions, but also do not mark them as static. Put the declaration in `waltham-private.h`. We only ever build a single `libwaltham.so` unlike with Wayland.

Am I missing something?